socketio/src/socket.rs

```diff
@@ -70,52 +70,85 @@ impl Socket {
-        let socket_packet = self.build_packet_for_payload(data, event, nsp, None)?;
+        self.emit_multi(nsp, event, vec![data])
+    }
+
+    /// Emits to certain event with given vector of data.
+    pub fn emit_multi(&self, nsp: &str, event: Event, data: Vec<Payload>) -> Result<()> {
+        let socket_packet = Self::build_packet_for_payloads(data, Some(event), nsp, None, false)?;
 
         self.send(socket_packet)
     }
 
     /// Returns a packet for a payload, could be used for bot binary and non binary
     /// events and acks. Convenance method.
     #[inline]
-    pub(crate) fn build_packet_for_payload<'a>(
-        &'a self,
-        payload: Payload,
-        event: Event,
-        nsp: &'a str,
+    pub(crate) fn build_packet_for_payloads(
+        payloads: Vec<Payload>,
+        event: Option<Event>,
+        nsp: &str,
         id: Option<i32>,
+        is_ack: bool,
     ) -> Result<Packet> {
-        match payload {
-            Payload::Binary(bin_data) => Ok(Packet::new(
-                if id.is_some() {
-                    PacketId::BinaryAck
-                } else {
-                    PacketId::BinaryEvent
-                },
-                nsp.to_owned(),
-                Some(serde_json::Value::String(event.into()).to_string()),
-                id,
-                1,
-                Some(vec![bin_data]),
-            )),
-            Payload::String(str_data) => {
-                let payload = if serde_json::from_str::<serde_json::Value>(&str_data).is_ok() {
-                    format!("[\"{}\",{}]", String::from(event), str_data)
-                } else {
-                    format!("[\"{}\",\"{}\"]", String::from(event), str_data)
-                };
-
-                Ok(Packet::new(
-                    PacketId::Event,
-                    nsp.to_owned(),
-                    Some(payload),
-                    id,
-                    0,
-                    None,
-                ))
+        let (data, attachments) = Self::encode_data(event, payloads);
+
+        let packet_type = match attachments.is_empty() {
+            true if is_ack => PacketId::Ack,
+            true => PacketId::Event,
+            false if is_ack => PacketId::BinaryAck,
+            false => PacketId::BinaryEvent,
+        };
+
+        Ok(Packet::new(
+            packet_type,
+            nsp.to_owned(),
+            Some(data),
+            id,
+            attachments.len() as u8,
+            Some(attachments),
+        ))
+    }
+
+    fn encode_data(event: Option<Event>, payloads: Vec<Payload>) -> (String, Vec<Bytes>) {
+        let mut attachments = vec![];
+        let mut data = "[".to_owned();
+
+        if let Some(event) = event {
+            data += &format!("\"{}\"", String::from(event));
+            if !payloads.is_empty() {
+                data += ","
+            }
+        }
+
+        Self::encode_payloads(&mut data, payloads, &mut attachments);
```

Comment:
nit: seems to be re-implementing a json serialization. Would storing in a temporary struct then use serde_json's to_string be simpler?